### PR TITLE
[Core] Make multiprocessing pool raise TypeError when provided a non-iterable in imap or imap_unordered

### LINKED ISSUE #24237: [Core][multiprocessing] Make Pool.imap() and Pool.imap_unordered() throw TypeError on non-iterable/iterator
### Description

Currently, the Ray implementations of `Pool.imap()` and `Pool.imap_unordered()` accept arguments of the type `(func, iterable)`, where `iterable` can be an *actual* iterable, an iterator, *or a single input*. In In the letter case, the `IMapIterator` constructor treats `iterable` as `[iterable]`. The code implementing this (since [this PR](https://github.com/ray-project/ray/pull/24117#), which is functionally identical to previous versions in this regard) is as follows:
```
        try:
            self._iterator = iter(iterable)
        except TypeError:
            # for compatibility with prior releases, encapsulate non-iterable in a list
            iterable = [iterable]
            self._iterator = iter(iterable)
```
This behaviour deviates from the core Python implementations of `map` and `multiprocessing.imap` and `multiprocessing.imap_unordered`, which all throw a `TypeError` when given a single value as `iterable`. 

As this potentially breaks existing code, this was proposed as an *issue* to be addressed in a future release here:
_Originally posted by @edoakes in https://github.com/ray-project/ray/issues/24117#issuecomment-1110214799_

### Use case

The aim is to have a consistent interface for:
* Python `map`
* `multiprocessing`: `imap` and `imap_unordered`
* `ray.util.multiprocessing`: `imap` and `imap_unordered`

### PR DESCRIPTION
**Pull request by cadedaniel**
### TL;DR
This PR modifies `ray.util.multiprocessing.Pool` so that `pool.imap` and `pool.imap_unordered` throw `TypeError` when given non-iterable inputs.

### More details
We want the API of `ray.util.multiprocessing.Pool` to closely resemble that of `multiprocessing.Pool`. One Ray user pointed out in https://github.com/ray-project/ray/issues/24237 that our `imap` and `imap_unordered` implementations differ slightly from `multiprocessing.Pool` in what arguments they accept. See the following usage:
```python3
#!/usr/bin/env python3

def main():
    import multiprocessing as mp
    with mp.Pool() as pool:
        assert_throws_on_noniterable(pool.imap)
        assert_throws_on_noniterable(pool.imap_unordered)

    import ray.util.multiprocessing as raymp
    with raymp.Pool() as pool:
        assert_throws_on_noniterable(pool.imap)
        assert_throws_on_noniterable(pool.imap_unordered)

def assert_throws_on_noniterable(method):
    iterable_input = [1, 2, 3]
    noniterable_input = 3

    assert len([_ for _ in method(f, iterable_input)]) == len(iterable_input)

    try:
        [_ for _ in method(f, noniterable_input)]
        assert False, f"expected TypeError for {method}"
    except TypeError:
        pass

def f(_):
    pass

if __name__ == '__main__':
    main()
```

With this PR, this code produces no exception. Without this PR, this fails:
```
Traceback (most recent call last):
  File "/data/ray/./cade.py", line 30, in <module>
    main()
  File "/data/ray/./cade.py", line 11, in main
    assert_throws_on_noniterable(pool.imap)
  File "/data/ray/./cade.py", line 22, in assert_throws_on_noniterable
    assert False, f"expected TypeError for {method}"
AssertionError: expected TypeError for <bound method Pool.imap of <ray.util.multiprocessing.pool.Pool object at 0x7f109b9ff640>>
```

Closes https://github.com/ray-project/ray/issues/24237

### CODE REVIEW

**Comment by cadedaniel** [comment]
We will wait until Ray 2.4 to merge this. In Ray 2.3 we will raise a deprecation warning, see PR https://github.com/ray-project/ray/pull/31845

File: python/ray/tests/test_multiprocessing.py [hunk 1]
```diff
@@ -508,6 +508,31 @@ def f(args):
...
+    with pytest.raises(TypeError, match="object is not iterable"):
+        pool.map(fn, non_iterable)
+
+    with pytest.raises(TypeError, match="object is not iterable"):
+        pool.map_async(fn, non_iterable)
+
+    with pytest.raises(TypeError, match="must be an iterable, not"):
+        pool.starmap(fn, [non_iterable])
+
+    with pytest.raises(TypeError, match="must be an iterable, not"):
```
Comment (edoakes) [inline python/ray/tests/test_multiprocessing.py:532]: where are these error messages coming from, the `iter` call? 

Not sure what it looks like, but should we instead wrap it and write a more user-friendly error message?

**Review by edoakes — APPROVED** [review]

**Comment by cadedaniel** [comment]
review or merge pls cc @scv119 @jjyao

**Comment by scv119** [comment]
nice!
